builder/docker/config.go

```diff
@@ -219,6 +222,11 @@ func (c *Config) Prepare(raws ...interface{}) ([]string, error) {
 			warnings = append(warnings, "when running a bootstrap build, the `pull` option is ignored and is replaced by `build.pull` (true by default)")
 			c.Pull = false
 		}
+
+		if c.Platform != "" {
+			warnings = append(warnings, "when running a bootstrap build, the `platform` option is ignored and is replaced by `build.platform`")
+		}
+		c.Platform = c.BuildConfig.Platform
```

Comment:
Understood, then setting `c.Platform` makes sense here.

I do have two questions there regarding usage:
1. If a configuration has `c.Platform` set but not `c.BuildConfig.Platform` (likely for users transitioning to using `build`), what should we do? Right now it seems that `c.Platform` will be set to empty, which could come as a surprise, so I would suggest setting `c.BuildConfig.Platform = c.Platform`, with a warning, or even erroring altogether in such a case, so users are aware that they need to update their configuration.
2. If both `c.Platform` and `c.BuildConfig.Platform` are set, right now we set `c.Platform = c.BuildConfig.Platform`, but I wonder if this should not be flat out an error if they conflict, or a warning like the one you added otherwise.

Reply:
Either approach (warning / error) works for me for both of those, with a slight preference for explicitly raising an error when these values conflict so that users know to update. If you concern, I can update the PR to raise an error in both cases.

Reply:
@lbajolet-hashicorp: when processing `c.BuildConfig`, would it work to simply raise an error when `c.Platform` is not null? The only time `c.BuildConfig` gets processed in config.go would seem to be when there is a `build` option provided... so unless we want to for some reason allow the user to specify identical `c.BuildConfig.Platform` and `c.Platform` values, which seems pointless, I believe checking whether the value is null would resolve both of your questions.

Reply:
Yeah, checking if `c.Platform` is not null and erroring if we are doing a bootstrapped build would be alright in my opinion.
I thought we could only error if both values are set and conflicting, but I would think it's not too far of a stretch to require users to update their template's `platform` attribute if they're bootstrapping.

Reply:
@jonathanmccormack, the more I think about this, the more I'm under the impression that we don't need to introduce `platform` to the `BuildConfig`, since both are conflicting and only one should be defined.
I would think keeping the current `platform` attribute the only one settable by users, and propagating that value to the `BuildConfig`, should be enough and would be simpler to implement.

What do you think?

Reply:
Fine by me, but may require a larger rewrite of the logic. I initially followed the example of `c.Pull` and `c.BuildConfig.Pull`, which also seem duplicative.  It would make sense to me for both `pull` and `platform` to be defined solely at the root configuration level (i.e., keep just `c.Pull` and `c.Platform`) and then keep the options available under `c.BuildConfig` limited just to those that are specific to a bootstrap build, like the `arguments` option from my earlier PR.

That said, it'll probably require a bit more refactoring to make sure the root configuration are available where they need to be in dockerfile_config.go (and wherever else `c.BuildConfig.Pull` is used, if improving the logic for both `pull` and `platform`). I can take a look at doing so if you'd prefer that approach.

Reply:
Yeah regarding `Pull` the main reason why we have both is because of `pull` being true by default for bootstrapped builds, and not for others. We can probably make them consistent and unified in a later release (tbh I think `pull` should be `true` by default, it seems more likely to me that users will want to pull their base image if it's not locally present), though this will be a breaking change, so that'll become a major bump for the plugin then.

Regarding `platform`, both should be harmonised from what I understand, so that makes it less useful to introduce both at top-level and for the build config, hence my suggestion to propagate `c.Platform` as `c.BuildConfig.Platform`, and not making the latter settable in a user configuration (the `mapstructure:"-"` tag should do the trick, or making the attribute private should also).

Reply:
Looks like specifying:
```go
Platform string `mapstructure-to-hcl2:",skip"`
```
...seems to suppress generating a line for `c.BuildConfig.Platform` in the docs while still disabling the ability for a user to set that value, but where specifying `c.Platform` works for a bootstrap build. @lbajolet-hashicorp, can you please confirm whether this should accomplish the goal? I am a bit out of my depth here.